src/core/comm/coll.cc

```diff
@@ -0,0 +1,443 @@
+/* Copyright 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <assert.h>
+#include <limits.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <atomic>
+#include <cstdlib>
+
+#ifndef LEGATE_USE_GASNET
+#include <stdint.h>
+#endif
+
+#include "coll.h"
+#include "legion.h"
+
+namespace legate {
+namespace comm {
+namespace coll {
+
+using namespace Legion;
+Logger log_coll("coll");
+
+#ifdef LEGATE_USE_GASNET
+
+enum CollTag : int {
+  BCAST_TAG     = 0,
+  GATHER_TAG    = 1,
+  ALLTOALL_TAG  = 2,
+  ALLTOALLV_TAG = 3,
+  MAX_TAG       = 10,
+};
+
+static std::vector<MPI_Comm> mpi_comms;
+#else  // undef LEGATE_USE_GASNET
+static std::vector<ThreadComm> thread_comms;
+#endif
+
+static int current_unique_id = 0;
+
+static bool coll_inited = false;
+
+// can be override by set the env LEGATE_MAX_COMMS
+static int MAX_NB_COMMS = 100;
+
+// functions start here
+int collCommCreate(CollComm global_comm,
+                   int global_comm_size,
+                   int global_rank,
+                   int unique_id,
+                   const int* mapping_table)
+{
+  global_comm->global_comm_size = global_comm_size;
+  global_comm->global_rank      = global_rank;
+  global_comm->status           = true;
+  global_comm->unique_id        = unique_id;
+#ifdef LEGATE_USE_GASNET
+  int mpi_rank, mpi_comm_size;
+  int *tag_ub, flag;
+  int compare_result;
+  MPI_Comm comm = mpi_comms[unique_id];
+  CHECK_MPI(MPI_Comm_compare(comm, MPI_COMM_WORLD, &compare_result));
+  assert(compare_result = MPI_CONGRUENT);
```

Comment:
needs an equality check instead of an assignment. fyi, it's a good coding practice to put constant values on the LHS of equality checks to avoid mistakes like this.

Reply:
nice catch! I will update the assert to put constant value on the left. Thank you